celltype_annotation/analysis/copykat-rms-exploration.Rmd

```diff
@@ -0,0 +1,224 @@
+        id.type = "E", # we have Ensembl gene ids, not the default gene symbols ("S")
+        sam.name = params$library_id, 
+        distance = x, 
+        plot.genes = FALSE,
+        n.cores = params$threads
+      )
+      
+      # save prediction to sce
+      colname <- glue::glue("copykat_{x}")
+      colData(sce)[[colname]] <- copykat_result$prediction$copykat.pred
+      
+      # export full result
+      output_file <- file.path(output_dir,
+                              glue::glue("{params$library}_copykat_{x}.rds"))
+      readr::write_rds(copykat_result, output_file)
+    }
+  )
+  
+  # finally, export SCE with all copykat predictions
+  readr::write_rds(sce, copykat_sce)
+} 
+```
+
+## copyKat results
+
+First, we'll look at quick tables for correspondance between submitter cell type annotations and `copyKat` inferences.
+For `copyKat`, "diploid" is normal and "aneuploid" is tumor.
+
+```{r}
+table(sce$celltype, sce$copykat_spearman)
+```
+
+```{r}
+table(sce$celltype, sce$copykat_pearson)
+```
+
+```{r}
+table(sce$celltype, sce$copykat_euclidian)
+```
+
+
+Next, we'll look at some classification metrics overall, using a custom function to calculate a confusion matrix.
+For this, we will _only consider_ cells which i) the submitter was able to classify, and ii) `copyKat` was able to classify.
+In other words, all `NA` values are removed from the data before calculations.
+
+
+
+```{r}
+# Custom function to calculate some confusion matrix quantities, including
+# accuracy, TPR, and FPR
+calculate_confusion <- function(df, test_column) {
+  # assumes the true column is `celltype_class`
+  df |>
+    # remove NA values all around
+    tidyr::drop_na(celltype_class) |>
+    dplyr::filter({{test_column}} != "not.defined") |>
+    # confusion matrix
+    dplyr::mutate(confusion = dplyr::case_when(
+      celltype_class == "Tumor" & {{test_column}} == "aneuploid"  ~ "TP",
+      celltype_class == "Tumor" & {{test_column}} == "diploid"    ~ "FN",
+      celltype_class == "Normal" & {{test_column}} == "aneuploid" ~ "FP",
+      celltype_class == "Normal" & {{test_column}} == "diploid"   ~ "TN",
+      TRUE ~ NA_character_,
+    )) |>
+    # just in case..
+    tidyr::drop_na(confusion) |>
+    dplyr::count(confusion) |>
+    tidyr::pivot_wider(
+      names_from = "confusion", 
+      values_from = "n"
+    ) |>
+    # calculate some metrics
+    dplyr::summarize(accuracy = (TP+TN)/(TP+TN+FP+FN), 
+                     tpr      = (TP)/(TP+FN),
+                     fpr      = (FP)/(FP+TN)) 
+}
+```
+
+
+```{r}
+# Create a data frame for performing calculations on
+copykat_df <- colData(sce) |>
+  as.data.frame() |>
+  dplyr::select(celltype, contains("copykat")) |>
+  dplyr::mutate(celltype_class = ifelse(
+    stringr::str_starts(celltype, "Tumor"), 
+    "Tumor", 
+    "Normal"
+  )) 
+
+# run confusion matrix function
+cat("============== SPEARMAN distance =====================")
+calculate_confusion(copykat_df, copykat_spearman)
```

Comment:
I think this is super helpful, but is there a way we could see the breakdown of accuracy per tumor/normal? I guess I'm trying to get an understanding of whether or not more normal cells are called as tumor here? Or vice versa?